samples/mpu/mem_domain_apis_test/src/main.c

```diff
@@ -0,0 +1,128 @@
+/* main.c - Memory Domain APIs demo */
+
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <misc/printk.h>
+
+/* size of stack area used by each thread */
+#define STACKSIZE 1024
+
+/* scheduling priority used by each thread */
+#define PRIORITY 7
+
+/* busy waiting time (in us) */
+#define BUSY_WAITING_TIME 3000000
+
+
+K_THREAD_STACK_ARRAY_DEFINE(app_stack, 3, STACKSIZE);
+
+struct k_thread app_thread_id[3];
+
+struct k_mem_domain app_domain[2];
+
+/* the start address of the MPU region needs to align with its size */
+u8_t __aligned(32) app0_buf[32];
+u8_t __aligned(32) app1_buf[32];
+
+struct k_mem_partition app0_parts[] = {
+	MEM_PARTITION_ENTRY((u32_t)app0_buf,
+			    sizeof(app0_buf),
+			    K_MEM_PARTITION_P_RW_U_RW |
+			    K_MEM_PARTITION_EXECUTE_NEVER),
+	MEM_PARTITION_ENTRY((u32_t)app1_buf,
+			    sizeof(app1_buf),
+			    K_MEM_PARTITION_P_RW_U_RO |
+			    K_MEM_PARTITION_EXECUTE_NEVER),
+};
+
+struct k_mem_partition app1_parts[] = {
+	MEM_PARTITION_ENTRY((u32_t)app1_buf,
+			    sizeof(app1_buf),
+			    K_MEM_PARTITION_P_RW_U_RW |
+			    K_MEM_PARTITION_EXECUTE_NEVER),
+	MEM_PARTITION_ENTRY((u32_t)app0_buf,
+			    sizeof(app0_buf),
+			    K_MEM_PARTITION_P_RW_U_RO |
+			    K_MEM_PARTITION_EXECUTE_NEVER),
+};
+
+/* application thread */
+void app_thread(void *dummy1, void *dummy2, void *dummy3)
+{
+	int id = (int) dummy1;
+	int count = 0;
+
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	printk("app thread %d (%p) starts.\n", id, &app_thread_id[id]);
+
+	/*
+	 * app thread 0 will be added into app0_domain.
+	 * app thread 1 will be added into app1_domain.
+	 * app thread 2 will be added into app2_domain.
+	 */
+	printk("app thread %d (%p) adds itself into app%d_domain.\n",
+		id, &app_thread_id[id], id%2);
+
+	k_mem_domain_add_thread(&app_domain[id%2],
+				&app_thread_id[id]);
+
+	k_busy_wait(BUSY_WAITING_TIME);
+
+	while (1) {
+		printk("app thread %d (%p) yields.\n", id, &app_thread_id[id]);
+
+		k_yield();
+
+		printk("app thread %d (%p) is back.\n", id, &app_thread_id[id]);
+
+		k_busy_wait(BUSY_WAITING_TIME);
+
+		if (count == 4 && id == 1) {
```

Comment:
The memory domain APIs is aimed for unprivileged threads to share data to the threads in the same memory and protect sensitive data from threads outside their domain. It is not only for improving the security but also useful for debugging (unexpected access would cause exception).
To get the best demonstration for memory domain APIs, we need to enable unprivileged threads
execution environment first. Since the related works for enabling unprivileged threads have not done, the alternative way I found to demonstrates these APIs is to use debug output to show the changes of MPU regions druing context switch. Observing the debug output, we could see kernel will load MPU regions of specific memory domain before threads in the domain are switched in. and disabe MPU regions if threads is not in any domain.
To use busy-waiting is that using other wating APIs (e.g. k_sleep) might cause idle thread being switch in, but I want the debug out always to show the messages related to app thread 0,1,2.